Magazine Viewer, Magazine by Publications, Bookmark Magazines

zachseidner1 opened the pull request:
## Overview

In this PR, the individual magazine viewer, the ability to view magazines by publication, and the ability to bookmark and view bookmarked magazines has been added, thus completing the magazines feature for Android. 

## Changes Made

- Implemented a PDF viewer for Jetpack compose with [Bouquet](https://github.com/GRizzi91/bouquet)
- Implemented loading bar for PDF download progress in individual viewer
- Magazine covers now load the first page of the magazine with mini-PDF viewers.
- Magazines can be bookmarked, shared, and shouted out. 
- Bookmarks page now has a section for viewing bookmarked magazines.
- Refactored publications page to use a lazy vertical grid to support both article and magazine viewing. 
- Change default selection from semester view to a "View all" button, which shows all available magazines. 

## Please note:

- Some features seem to only work on the prod server, such as loading featured magazines or loading all magazines. This _should_ not be an issue when the app is shipped to the Play Store.
- Following publications can be a bit inconsistent depending on what server you are on, again this should not be an issue when the app is shipped.
- On emulators, the PDF viewer can be a bit laggy. I implemented everything I could using the tools available in the library to optimize the experience. On the Android test phone, this behavior is **not** shown.

## Next Steps

Magazines are finished! Yay me. Next step will be implementing the community board, and potentially adding some smaller features and bug fixes to bring feature parity. Also will probably be implementing dedicated cover images instead of loading a lot of mini-PDF viewers which will optimize the experience.

## Screenshots & Videos

<!-- If you are making any changes to UI, you should use this section. -->

<details>
  <summary>Viewing a Magazine with the PDF Viewer</summary>
<img src="https://user-images.githubusercontent.com/58796478/228035904-2e5551a0-0201-41ad-a382-f5c7f96d7567.png" 
width=300/> 
</details>


<details>
  <summary>Viewing Bookmarked Magazines</summary>
<img src="https://user-images.githubusercontent.com/58796478/228036314-f9e6fecf-a46e-4b94-b009-ef5c0203a7ad.png" 
width=300/> 
</details>


<details>
  <summary>Viewing Magazines By Publication</summary>
<img src="https://user-images.githubusercontent.com/58796478/228036525-0d60626d-7edf-4a60-b8c8-18fcf6dea79f.png" 
width=300/> 
</details>

zachseidner1 committed:
Implemented PDFViewer, bottom bar, and visuals for top bar
- PDF Viewer has been implemented and PDFs can now be read using it. Uses the bouquet library.
- Bottom bar has been implemented, including shoutouts and bookmarking.
- Top bar aesthetics have been implemented, but not yet functional

zachseidner1 committed:
Implemented top bar, bookmarked magazines viewer
- Top bar was implemented with functional back button
- Magazines viewer on bookmark page was implemented, and bookmark view model was updated.
- For some reason, following publications is currently broken which I need to fix.

zachseidner1 committed:
Merge pull request #1 from cuappdev/main

Merge pull request #24 from zachseidner1/main

zachseidner1 committed:
Merge pull request #27 from zachseidner1/Zach/View-Magazine

Push to main repo

ckdesir committed:
Add missing social media logos

ckdesir committed:
Change gradle to switch to dev server when debugging

ckdesir committed:
Change to itemIndexed for smoother scrolling

zachseidner1 committed:
Fixed following of publications

zachseidner1 committed:
Networking for loading magazines by publication
- Implemented basic networks for finding magazines by publication (untested)
- Need to implement way to view these magazines when looking at a publication

zachseidner1 committed:
Magazines by Publication
- Added the ability to view magazines while browsing by publication.
- Refactored publication screen to lazy vertical grid to support global magazine scrolling.

zachseidner1 committed:
Small Magazine Optimizations
- Changed View Models to query for magazines whether or not the articles were able to load
- Fixed position of loading icons in bookmarks page.

zachseidner1 committed:
Center bottom and top bars of individual magazine viewer

zachseidner1 committed:
View ALl Magazines Prod Server Fix
- Previously, if selected a different semester and then tried to reselct 'View all', no magazines would load.
- This functionality now works and the proper query is made when 'View all' is selected.

Corfish123 commented:
I tried running the app and there were some issues I noted. First, when I am in the magazine tab, and I am under View All, it is constantly loading and nothing ever shows up. From the log statements it seems like you don't actually do any calling as this was the network call: {"data":{"getMagazinesBySemester":[]}}. As a result nothing is loading

2. When you are on the magazine screen, and try scrolling, you are able to scroll, however, it is only possible if you are scrolling when your first touch is between the preview of the magazine. If you try to scroll when you finger is on the preview of the magazine, it does not scroll. They are able to scroll while they are still loading with the shimmer effect but once they load in, you can no longer press there to scroll. As a result, this causes a lot of dead zones where you can not scroll. It seems the issue might be the pdf previewer composable that appears. 
![image](https://user-images.githubusercontent.com/80220215/228405355-0017b62d-66ce-4551-bcb7-1c81adb8bc68.png)
So basically all the highlighted areas you can scroll from there. 

3. You also mentioned that you made tabs for the publications here
![image](https://user-images.githubusercontent.com/80220215/228405153-dba57464-4273-4935-b874-e3cfe2ec84f5.png)
However, I do not see the tab appear in this PR.

zachseidner1 committed:
Fixed debug endpoint to DEV_ENDPOINT

Corfish123 commented on app/src/main/java/com/cornellappdev/android/volume/ui/screens/HomeScreen.kt:
```diff
@@ -87,36 +88,55 @@ fun HomeScreen(
                             }
                         }
+                        showPageBreak = true
+//                            Box(modifier = Modifier.padding(top = 10.dp)) {
```
Delete this when pushing the main

Corfish123 commented on app/src/main/java/com/cornellappdev/android/volume/ui/screens/IndividualMagazineScreen.kt:
```diff
@@ -0,0 +1,296 @@
+import com.rizzi.bouquet.rememberHorizontalPdfReaderState
+
+private const val TAG = "IndividualMagazineScreen"
+
```
Add a comment here

Corfish123 commented on app/src/main/java/com/cornellappdev/android/volume/ui/screens/IndividualPublicationScreen.kt:
```diff
@@ -1,46 +1,55 @@
-                    ) {
-                        CircularProgressIndicator(color = VolumeOrange)
-                    }
+                    Log.d(TAG, "IndividualPublicationScreen: Loading state reached")
```
delete these log statements

Corfish123 commented on app/src/main/java/com/cornellappdev/android/volume/ui/screens/MagazinesScreen.kt:
```diff
@@ -166,31 +179,38 @@ fun MagazinesScreen(
+{
     when (val magazinesState =  magazineUiState.featuredMagazinesState) {
         MagazinesRetrievalState.Loading -> {
+            Log.d(TAG, "FillFeaturedMagazinesRow: MAGAZINE LOADING")
```
delete these logs as well

Corfish123 commented on app/src/main/java/com/cornellappdev/android/volume/ui/viewmodels/IndividualPublicationViewModel.kt:
```diff
@@ -59,18 +60,22 @@ class IndividualPublicationViewModel @Inject constructor(
 
     private fun queryPublication() = viewModelScope.launch {
         try {
+            Log.d(TAG, "queryPublication: No null pointers")
```
delete these log statements

Corfish123 commented on app/src/main/java/com/cornellappdev/android/volume/ui/viewmodels/IndividualPublicationViewModel.kt:
```diff
@@ -79,16 +84,34 @@ class IndividualPublicationViewModel @Inject constructor(
+    }
+
+    private fun queryMagazinesByPublication() = viewModelScope.launch {
+        Log.d(TAG, "queryMagazinesByPublication: Querying magazines")
```
delete this

Corfish123 reviewed (commented):
Keep up the good work!! Just remember to delete all of your testing log statements before pushing it

zachseidner1 committed:
Documentation Improvement
- Deleted unnecessary Log.d statements
- Added documentation to IndividualMagazineScreen
- Removed commented out code

Corfish123 reviewed (approved):
Good stuff!! Keep up the good work
